audacity : update and rewrite

RJVB opened the pull request:
#### Description

This upgrades port:audacity to the current stable release 2.4.2 and
port:audacity-devel to 2.4.2.790 (aka 3.0.0 beta). The main difference
between these versions is that the latter introduces and uses a new,
sqlite3-based file format (.aup3) which the older version cannot read
(and the newer version cannot export to the older .aup format).

A new subport is introduced, audacity-manual, which installs a local
copy of the online manual inside the app bundle where the application
will find it.

Several behind-the-scenes changes:
- the build is now handled through cmake and ninja
- port:audacity can depend on wxWidgets-3.0 and wxWidgets-3.2 via variants,
  port:audacity-devel will only build against wxWidgets-3.2 (by choice).
- the preferred way to build and run Audacity is against a patched version
  of wxWidgets 3.1.3 which contains fixes not included upstreams plus some
  new functionality (including accessibility support on Mac). The cmake
  build system has logic to grab the patched sources or use a locally
  provided copy, build them, and embed the required libraries in the app
  bundle. This has little effect on the size (more on the build time...)
  Therefore, the default build of the ports now uses this solution to
  provide the best possible user experience. Users who insist on runnnig
  against a "system" wxWidgets can set the +wxsystem variant, in which
  case the choice(s) outlined above become available.

Fixes: 5693


###### Tested on
macOS 10.9.5
clang-mp-6.0

###### Verification <!-- (delete not applicable items) -->
Have you

- [X ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ X] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ X] tried a full install with `sudo port -vst install`?
- [ X] tested basic functionality of all binary files?

ra1nb0w commented:
thank you @RJVB . Can you update to 3.0 and then I will give a try..

cooljeanius commented:
What are the plans for this going forward given the number of people forking Audacity due to their recent changes to start adding telemetry and stuff?

mojca commented:
> What are the plans for this going forward given the number of people forking Audacity due to their recent changes to start adding telemetry and stuff?

My suggestion would be to stick with the official release and just turn telemetry off for everyone as the first step. Then we can see how things unroll with time.

pmetzger commented:
The code now also demands that users be over 13 IIRC because of the telemetry. I think at the very least we should also package Tenacity or one of the others?

RJVB commented:
I've been exchanging with the guy who set up an "aud-free" repository, currently with a nodejs recipe but hopefully also with patchfiles for release versions. I'll take that back up and see what can be done when I find the time for it.

Either way, the version currently in the PR is not yet concerned by all this and IMVHO works perfectly fine.

reneeotten commented:
so let's do something with this PR - either merge it (after resolving the conflicts) or close it. I'm not using this port so have no real preference, just trying to clear out the PR queue ;) 

Are there any concerns about merging this, from my reading it looks like the comments are all related to changes in upcoming versions but not related to *this* specific update, correct?

pmetzger commented:
I have no problem with this being merged (provided it is fixed to work.)

RJVB commented:
I've been meaning to get back to this. Don't hold your breath too long but I'll get around to it.

pmetzger commented:
Rather than "not holding our breath" why don't we just commit this or close it?

RJVB commented:
I've stated my intentions, I can't speak for the (other) persons in `we`...

cooljeanius commented:
> Rather than "not holding our breath" why don't we just commit this or close it?

We can't just commit it because there's a merge conflict, and we shouldn't close it, because RJVB intends to get back to it.

RJVB commented:
Oh, I can always open a new PR, if I can't reopen this one.

ssight commented:
Hi @RJVB,

Sorry for completely fogetting about free-aud; I have been exceedingly busy these past few months, and the entire project seemed to be lost to the back of my mind.

With the community largely having accepted Tenacity as the primary, truly open-source, edition of Audacity, I see little reason in the continuation of free-aud. If I were you, I would invest my time in Tenacity, as it seems far more supported and stable, via the open source community, than I will ever be able to achieve on my own.

However, if you truly believe free-aud is the way to go, I will be more than happy to continue to work on the project and get everything up and running once again.

Sorry for the abcense and thanks once again for keeping in touch.

RJVB commented:
Hello @ssight,

>However, if you truly believe free-aud is the way to go, I will be more than happy to continue to work on the project and get everything up and running once again.

If it were just for me, I would indeed simply jump to Tenacity (or another fork, if I felt I had a reason for that). Here, we're talking about packaging, so there would be something to say for patching the unwanted bits out of the official code. Then again, there would also be something to say to go with Tenacity if they have an actual community that is open for contributions that are never even proposed to Audacity, more interested in addressing issues, etc. We'll see what feedback I get from the MacPorts powers who be...!

I've only just started to look at Tenacity but already noticed that the source tree is quite different from the 3.0.0 beta version I was using until now (#463058e20cf804d06bf1abe3ac629f0eba0e6bdd). It could be that this wouldn't be any different with Audacity, of course.

RJVB commented:
Hi again,

>>However, if you truly believe free-aud is the way to go, I will be more than happy to continue to work on the project and get everything up and running once again.
>
>If it were just for me, I would indeed simply jump to Tenacity (or another fork, if I felt I had a reason for that). Here, we're talking about packaging, so there would be something 

I see a few good reasons not to use Tenacity:

- they now target 10.12 as the minimum Mac OS version. While this may not yet reflect the current state of things it could very well become so quickly.
- all "vendored dependencies" were removed according to https://github.com/tenacityteam/tenacity/pull/228, moving to using vcpkg to obtain those dependencies (merged partially via https://github.com/tenacityteam/tenacity/commit/ebd31bb3762f5e5c11b9a8fe5e3fb8f059c10ebe that touched just 5839 files :-O ).

The latter should not be a problem in itself if vcpkg can be clever enough not to install dependencies that are already available through MacPorts; we have always used a part of those vendored dependencies because either they aren't in MacPorts (vamp, sbsms), or there is some kind of incompatibility (apparently; portmidi, portaudio). In my last implementation I even gave up trying to build against a stock wxWidgets but instead use "their" version.

I see other things going on that clearly aim at building "proper", standalone app bundles on Mac, and I can feel that it's going to become increasingly more work to keep Tenacity building "the MacPorts way". Also, according to https://github.com/tenacityteam/tenacity/issues/506 we would lose LV2 plugin support in the MacPorts build unless we add a port for suil .

All in all this is way more involved than I thought it'd be and then I'm prepared to handle at the moment (if ever...), so I hope @ssight can indeed find the time at least to fix free-aud so that it (= a clean checkout of the repo, built according to the docs) runs again on the Audacity version he last checked it on. I can help with things like relocating files that were renamed or checking for other/moved occurrences of "malicious" code.

R.

Be-ing commented:
Building Tenacity with vcpkg is not a requirement nor is it intended to be used when packaging Tenacity in other package managers. It is enabled by default on macOS; just pass `-D VCPKG=OFF` to cmake to disable it.

Be-ing commented:
I find it odd that you want to use vcpkg inside another package manager, but that's up to you. :shrug: Other packagers got concerned that they would have to use vcpkg (which they don't).

> I don't know why you'd use such a modern dialect

Audacity is using C++17 also. I do not share your concern with using modern standards.

RJVB commented:
I've updated the audacity-devel subport to 3.0.2.129, which is the newest version (AFAICT!) that shouldn't yet have any of the commits after the controversial project takeover. I suggest we stick to that version until we decide what fork to track in the future and/or until I work something out with @ssight (= an additional "cleansing" patch).

I've stuck to my earlier proposition to provide v2.4.2 as the main port, but we could swap the subport logic: `audacity` -> `audacity-legacy` and `audacity-devel` -> `audacity` (yes, I feel confident providing in-between releases code as a main rather than as a -devel port).

pmetzger commented:
@RJVB It's up to you I think. Don't know what everyone else believes though.

RJVB commented:
Hi @ssight, just thought I'd check in and see if you had done anything on this project, but it seems to have been taken down completely (not just archived)?

n0toose commented:
> Hi @ssight, just thought I'd check in and see if you had done anything on this project, but it seems to have been taken down completely (not just archived)?

I don't think I get the context 100%, but if you're referring to Tenacity, it's under new management: https://codeberg.org/tenacityteam/tenacity

RJVB commented:
ssight's "free-aud" was a project that aimed to remove the dubious aspects from Audacity via text processing instead of patches.

Either way, I've been going over the source and configure options of the latest Audacity code and as far as I can tell it is (now) possible to deactivate the phoning-home features (which are apparently which require changes to the privacy policy), crash-report sending and 1 or 2 other features we don't want/need in Macports.

Hopefully I'll be able to clean up the portfile and make it build a current version. Low priority business for me though.
